swbemservices.go

```diff
@@ -0,0 +1,287 @@
+// +build windows
+
+package wmi
+
+import (
+	"fmt"
+	"github.com/go-ole/go-ole"
+	"github.com/go-ole/go-ole/oleutil"
+	"reflect"
+	"runtime"
+)
+
+// SWbemServices is used to access wmi. See https://msdn.microsoft.com/en-us/library/aa393719(v=vs.85).aspx
+type SWbemServices struct {
+	//TODO: track namespace. Not sure if we can re connect to a different namespace using the same instance
+	cWMIClient            *Client //This could also be an embedded struct, but then we would need to branch on Client vs SWbemServices in the Query method
+	sWbemLocatorIUnknown  *ole.IUnknown
+	sWbemLocatorIDispatch *ole.IDispatch
+	queries               chan *queryRequest
+	closeError            chan error
+}
+
+type queryRequest struct {
+	query    string
+	dst      interface{}
+	args     interface{}
+	finished chan error
+}
+
+// InitializeSWbemServices will return a new SWbemServices object that can be used to query WMI
+func InitializeSWbemServices(c *Client, connectServerArgs ...interface{}) (*SWbemServices, error) {
+	fmt.Println("InitializeSWbemServices: Starting")
+	s := new(SWbemServices)
+	s.cWMIClient = c
+	s.queries = make(chan *queryRequest)
+	initError := make(chan error)
+	go s.process(initError)
+
+WaitForProcessSignal:
+	for {
+		select {
+		case err, ok := <-initError:
+			if ok {
+				return nil, err //Send error to caller
+			}
+			break WaitForProcessSignal
+		default:
+			//Wait for process method to return an error or signal it is ready (by closing channel)
+		}
+	}
+	fmt.Println("InitializeSWbemServices: Finished")
+	return s, nil
+}
+
+// Close will clear and release all of the SWbemServices resources
+func (s *SWbemServices) Close() error {
+	fmt.Println("Close: sending close request")
+	var result error
+	ce := make(chan error)
+	s.closeError = ce //Race condition if multiple callers to close. May need to lock here
```

Comment:
I agree. a lock for close would be good. Make sure to check queries after you get the lock as well. It could be closed / nil.